zenith_utils/src/pq_proto.rs

```diff
@@ -798,3 +813,156 @@ impl<'a> BeMessage<'a> {
         Ok(())
     }
 }
+
+// Zenith extension of postgres replication protocol
+// See ZENITH_STATUS_UPDATE_TAG_BYTE
+#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
+pub struct ZenithFeedback {
+    // Last known size of the timeline. Used to enforce timeline size limit.
+    pub current_instance_size: u64,
+    // Parts of StandbyStatusUpdate we resend to compute via safekeeper
+    pub ps_writelsn: u64,
+    pub ps_applylsn: u64,
+    pub ps_flushlsn: u64,
+    pub ps_replytime: SystemTime,
+}
+pub const ZENITH_FEEDBACK_FIELDS_NUMBER: u8 = 5;
+
+impl ZenithFeedback {
+    pub fn empty() -> ZenithFeedback {
+        ZenithFeedback {
+            current_instance_size: 0,
+            ps_writelsn: 0,
+            ps_applylsn: 0,
+            ps_flushlsn: 0,
+            ps_replytime: SystemTime::now(),
+        }
+    }
+
+    // use custom serialization to support protocol extensibility
+    pub fn serialize(&self, buf: &mut BytesMut) -> Result<()> {
+        buf.put_u8(ZENITH_FEEDBACK_FIELDS_NUMBER); // # of keys
+        write_cstr(&Bytes::from("current_instance_size"), buf)?;
+        buf.put_i32(8);
+        buf.put_u64(self.current_instance_size);
+
+        write_cstr(&Bytes::from("ps_writelsn"), buf)?;
+        buf.put_i32(8);
+        buf.put_u64(self.ps_writelsn);
+        write_cstr(&Bytes::from("ps_flushlsn"), buf)?;
+        buf.put_i32(8);
+        buf.put_u64(self.ps_flushlsn);
+        write_cstr(&Bytes::from("ps_applylsn"), buf)?;
+        buf.put_i32(8);
+        buf.put_u64(self.ps_applylsn);
+
+        let timestamp = match self.ps_replytime.duration_since(*PG_EPOCH) {
+            Ok(d) => d.as_micros() as i64,
+            Err(e) => -(e.duration().as_micros() as i64),
```

Comment:
Does it mean that we can be before PG_EPOCH? I think in that case we can bubble up the error but I cant imagine how can we actually trigger this case

Reply:
It is a copy-paste it from rust-postgres `standby_status_update()` implementation.
But, I don't see how can we get here, either.
So I've changed it to throwing error.